src/client/signatures/rc_crypto_verifier.rs

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+use {
+    super::x509::fetch_public_key,
+    super::{Collection, SignatureError, Verification},
+    canonical_json,
+    log::debug,
+    rc_crypto::signature,
+    serde_json::json,
+};
+
+pub struct RcCryptoVerifier {}
+
+impl RcCryptoVerifier {}
+
+impl Verification for RcCryptoVerifier {
+    fn verify(&self, collection: &Collection) -> Result<(), SignatureError> {
+        debug!("Verifying using x509-parser and rc_crypto");
+
+        // Fetch certificate PEM (public key).
+        let x5u = collection.metadata["signature"]["x5u"].as_str().ok_or(
+            SignatureError::InvalidSignature {
+                name: "x5u field not present in signature".to_owned(),
+            },
+        )?;
+
+        let public_key_bytes = fetch_public_key(&x5u)?;
+
+        // Get public key from certificates
+        let public_key =
+            signature::UnparsedPublicKey::new(&signature::ECDSA_P384_SHA384, &public_key_bytes);
+
+        // Instantiate signature
+        let b64_signature = collection.metadata["signature"]["signature"]
+            .as_str()
+            .unwrap_or("");
+        let signature_bytes = base64::decode_config(&b64_signature, base64::URL_SAFE)?;
+
+        // Serialized data.
+        let mut sorted_records = collection.records.to_vec();
+        sorted_records.sort_by(|a, b| (a["id"]).to_string().cmp(&b["id"].to_string()));
```

Comment:
This can be simplified using `sort_by_key`, e.g.

    sorted_records.sort_by_cached_key(|a| a["id"].to_string());

Not that `to_string()` actually performs a JSON serialization, so if the ids are strings, you sort quoted strings, possibly with unicode escapes. It's unclear whether these are the desired semantics.

The version you have performs the JSON serialization for every comparison. The version I proposed only performs it once for each collection.

Reply:
Awesome!

We are sure that `id`s are always present and will never have unicode, so I used `sorted_records.sort_by_cached_key(|a| a["id"].as_str().unwrap().to_owned());`